test/unittest/tests/block/test_block_api_write.cpp

```diff
@@ -0,0 +1,213 @@
+    REQUIRE(bm->addr_invalid(bm, session->get_wt_session_impl(), cookie, cookie_size) == 0);
+
+    // Test that the write buffer doesn't get modified after performing write.
+    CHECK(
+      memcmp(expected_str.c_str(), WT_BLOCK_HEADER_BYTE(write_buf->mem), expected_str.size()) == 0);
+
+    // Test that the write buffer was correctly written via performing a read.
+    WT_ITEM read_buf;
+    WT_CLEAR(read_buf);
+    REQUIRE(bm->read(bm, session->get_wt_session_impl(), &read_buf, cookie, cookie_size) == 0);
+    WT_BLOCK_HEADER *blk = reinterpret_cast<WT_BLOCK_HEADER *>(WT_BLOCK_HEADER_REF(write_buf->mem));
+    blk->checksum = 0;
+    CHECK(memcmp(write_buf->mem, read_buf.mem, write_buf->size) == 0);
+
+    // Test block header members.
+    CHECK(blk->disk_size == write_buf->memsize);
+    if (data_checksum)
+        CHECK(blk->flags == WT_BLOCK_DATA_CKSUM);
+    else
+        CHECK(blk->flags == 0);
+    __wt_buf_free(nullptr, &read_buf);
+}
+
+// Test that previous write buffers are still present in the block manager.
+void
+test_validate_cookies(WT_BM *bm, std::shared_ptr<mock_session> session,
+  const std::vector<std::pair<std::array<uint8_t, WT_BTREE_MAX_ADDR_COOKIE>, size_t>> &cookies,
+  const std::vector<std::string> &expected_strings)
+{
+    for (int i = 0; i < cookies.size(); i++) {
+        const auto &cookie = cookies.at(i);
+
+        // Make sure that the cookie is valid still.
+        REQUIRE(bm->addr_invalid(
+                  bm, session->get_wt_session_impl(), cookie.first.data(), cookie.second) == 0);
+
+        WT_ITEM read_buf;
+        WT_CLEAR(read_buf);
+        REQUIRE(bm->read(bm, session->get_wt_session_impl(), &read_buf, cookie.first.data(),
+                  cookie.second) == 0);
+
+        // The data should match the expected string.
+        const auto &str = expected_strings.at(i);
+        CHECK(memcmp(str.data(), WT_BLOCK_HEADER_BYTE(read_buf.mem), str.size()) == 0);
+        __wt_buf_free(nullptr, &read_buf);
+    }
+}
+
+void
+create_write_buffer(WT_BM *bm, std::shared_ptr<mock_session> session, std::string contents,
+  WT_ITEM *buf, size_t buf_memsize)
+{
+    // Fetch write buffer size from block manager.
+    REQUIRE(bm->write_size(bm, session->get_wt_session_impl(), &buf_memsize) == 0);
+    test_and_validate_write_size(bm, session, buf_memsize);
+
+    // Initialize the buffer with aligned size.
+    F_SET(buf, WT_ITEM_ALIGNED);
+    REQUIRE(__wt_buf_initsize(session->get_wt_session_impl(), buf, buf_memsize) == 0);
+
+    // Copy content string into the buffer.
+    // Following the architecture guide, it seems that the block manager expects a block header.
+    // I have tried to mimic that here.
+    REQUIRE(__wt_buf_grow_worker(session->get_wt_session_impl(), buf, buf->size) == 0);
+    memcpy(WT_BLOCK_HEADER_BYTE(buf->mem), contents.c_str(), contents.length());
+    // REQUIRE(__wt_buf_set(session->get_wt_session_impl(), buf, contents.c_str(),
```

Comment:
Need some eyes here, i can't use __wt_buf_set anymore because I need to memcpy at an offset.